File: javascript/section05/exercises/exercise02.js
// Notes:
// The arrays aren't sorted
// The input is two arrays there is a logic to scan them and see if there is any common elements and the result should return true or false
// The return means logging true or false or return a boolean value?
// They will always be of different sizes?

const array1 = ["a", "b", "c", "x"];
const array2 = ["z", "y", "i"];

const arrayA = ["a", "b", "c", "x"];
const arrayB = ["z", "y", "x"];

//Always think of different inputs

//What if we had two same letters in the same array? Q1
const arrayC = ["a", "a", "c", "x"];
const arrayD = ["z", "y", "x"];

//What if we had same letters in array and an equal letter in the other array? Q2
const arrayE = ["a", "a", "c", "x"];
const arrayF = ["z", "a", "x"];

//What if we had numbers? Q3
const arrayG = ["a", 1, "c", "x"];
const arrayH = ["z", 1, "x"];

//What if we had arrays as elements in the arrays? Q4
const arrayI = ["a", [], "c", "x"];
const arrayJ = ["z", [], "x"];

//What if we had null as elements in the array? Q5
const arrayK = ["a", null, "c", "x"];
const arrayL = ["z", null, "x"];

/* What if we didn't receive a second array? What would happen to the function? Q6 */
const arrayM = ["a", "b", "c", "x"];

//Bruteforce approach
const checkArraysIfThereIsCommonElement = (arrayA, arrayB) => { // O(n^2) = quadratic if the arrays are the same size
    //If the inputs sizes are different they look more like o(n * m)
    for (let i = 0; i < arrayA.length; i++) {
        for (let j = 0; j < arrayB.length; j++) {
            if (arrayA[i] == arrayB[j]) {
                console.log(true);
                return true;
            }
        }
    }
    return false;
};

//Always make the solution error free as possible

/*
Ask questions like: can we assume always two params?
I'll always be receiving the same data type as parameters?

Ex: func (array1:string[], array2:number[]) or (array1:string[], array2:number)

Make suggestions on how to improve it like doing  checks and etc...


Tell the interviewer about the weakpoints in your method like:

    in this function only numbers, strings and booleans can be used correctly here

    If we used non literal values this might not work like a javascript expression

    And also this code could be a little bit more readable

*/
const checkCommonItensInArrayNplusMSolution = (arrayA, arrayB) => { // O(n + m)
    let set = new Set();

    for (let i = 0; i < arrayA.length; i++) {
        set.add(arrayA[i]);
    }

    for (let i = 0; i < arrayB.length; i++) {
        if (set.has(arrayB[i])) {
            console.log("It has");
            return true;
        }
    }

    return false;

};

/*
    One of the most common patterns when it comes to nested for loops 
    is that you might be able to use hash tables in javascript they are called objects 
*/

//Always modularize your code breaking it into smaller functions
const mapArrayToObject = (array) => {
    let object = {};
    array.forEach(item => {
        if (!object[item]) 
            object[item] = true;
    });
    return object;
}

const neagoieSolution = (array1, array2) => {
    let map = mapArrayToObject(array1);
    //Loop through the second array and check if item in second array exists on created object
    for (let j = 0; j < array2.length; j++) {
        if (map[array2[j]]) {
            return true;
        }
    }
    return false;
}


const neagoieCleanerSolution = (array1,array2) => {
    return array1.some(item => array2.includes(item));
}
/*
    In this solution we used built-in methods in javascript having a nicer solution
*/

/*

    Q1 - It wouldn't add the a a second time because were checking it and it would return true because of x 

    Q3  & Q5 - This is a language specific question but the answer in javascript is yes because it would treat them as literals
    
    Q4 - In JavaScript, comparing arrays or objects checks reference equality (not value equality),
    so it would treat `[]` as unique, even if both arrays contain empty arrays. In `["a", [], "c", "x"]` and `["z", [], "x"]`,
    it would return `true` because of "x", but the empty arrays would not be considered equal.

    Q6 
    The function would break because there's nothing to iterate over

*/

console.log(neagoieSolution(arrayA,arrayB));

File: java/DSA/src/main/java/section05/exercises/Exercise02.java
package section05.exercises;

import java.util.HashSet;
import java.util.Set;
import java.util.Arrays;

public class Exercise02 {

    // Notes:
    // The arrays aren't sorted.
    // The input is two arrays. The logic scans them and checks if there are any common elements.
    // The result should return true or false.
    // Arrays may be of different sizes.

    public static void main(String[] args) {
        String[] array1 = {"a", "b", "c", "x"};
        String[] array2 = {"z", "y", "i"};

        String[] arrayA = {"a", "b", "c", "x"};
        String[] arrayB = {"z", "y", "x"};

        System.out.println(neagoieCleanerSolution(arrayA, arrayB));  // Expected output: true
    }

    // Brute force approach - O(n^2) for arrays of the same size
    // If arrays have different sizes, complexity is O(n * m)
    public static boolean checkArraysIfThereIsCommonElement(String[] arrayA, String[] arrayB) {
        for (int i = 0; i < arrayA.length; i++) {
            for (int j = 0; j < arrayB.length; j++) {
                if (arrayA[i].equals(arrayB[j])) {
                    System.out.println(true);
                    return true;
                }
            }
        }
        return false;
    }

    // Always aim to make the solution error-free as possible

    /*
    Ask questions such as:
    - Can we always assume two parameters?
    - Will we always receive arrays of the same data type?

    For example:
    - `method(String[] array1, Integer[] array2)` or `method(String[] array1, int[] array2)`

    Make suggestions for improvement, such as adding null checks or size validation.
    */

    public static boolean checkCommonItemsInArrayNPlusMSolution(String[] arrayA, String[] arrayB) { // O(n + m)
        Set<String> set = new HashSet<>();

        // Populate the set with the elements of arrayA
        for (String item : arrayA) {
            set.add(item);
        }

        // Check if any element from arrayB exists in the set
        for (String item : arrayB) {
            if (set.contains(item)) {
                System.out.println("It has");
                return true;
            }
        }

        return false;
    }

    /*
        In Java, we can use HashSets for O(n) lookups.
        HashMaps can also be useful for storing key-value pairs, but for this solution, a Set is sufficient.
    */

    // Helper method to map array elements into a Set
    // This prevents adding duplicates, which is handled automatically by HashSet in Java
    public static Set<String> mapArrayToSet(String[] array) {
        Set<String> set = new HashSet<>(Arrays.asList(array));
        return set;
    }

    // Neagoie solution using a HashSet
    public static boolean neagoieSolution(String[] array1, String[] array2) {
        Set<String> set = mapArrayToSet(array1);

        // Loop through the second array and check if any element exists in the set
        for (String item : array2) {
            if (set.contains(item)) {
                return true;
            }
        }

        return false;
    }

    // Neagoie cleaner solution using Java's built-in stream functions
    public static boolean neagoieCleanerSolution(String[] array1, String[] array2) {
        return Arrays.stream(array1).anyMatch(item -> Arrays.asList(array2).contains(item));
    }

    /*
        This cleaner solution uses Java's `Stream` API to implement a concise, readable solution.
        It behaves similarly to the `.some()` and `.includes()` functions in JavaScript.

        Q1 - If we had two identical letters in the same array, only one would be added to the set since sets prevent duplicates.
        Q3 & Q5 - In Java, primitive types like `int` would need to be handled separately if arrays were mixed types.
        Q4 - Java compares object references for arrays, meaning `[]` would be treated as unique and not equal to other empty arrays.
        Q6 - If we didn't receive a second array, the function would throw a `NullPointerException`. Adding null checks would prevent this.
    */
}
